### PR TITLE
Bump terraform version to 1.5.1

### PR DESCRIPTION
**Pull request by TylerHendrickson**
## Description

This PR mostly consists of chore updates to Terraform dependency versions (which are being neglected by Dependabot for some reason):
- Terraform: 1.3.9 ➡️  1.5.1
- AWS provider: 4.55.0 ➡️  5.4.0
  - Note: this is a major version upgrade, although it doesn't seem to be causing any unexpected/breaking changes in this repo.
- Datadog provider: 3.23.0 ➡️  3.26.0
- HTTP provider: 3.2.1 ➡️  3.4.0 

I also included a "developer goodness" enhancement by way of adding a new `prebuild-lambda` task definition to our Taskfile. This is intended to help address a pain point reported by developers where running `task build` or multiple concurrent `task build-<LambdaName>` commands for the first time (or whenever the go cache is sufficiently stale) incurs a noticeable performance hit to the system. In particular, this pain is felt when running `terraform apply`, which runs `task build-<LambdaName>` more or less concurrently across all Lambda functions when generating source artifacts. Because the artifacts are built in isolation, the Go compiler processes can take on a lot of duplicate work – for example, when multiple Lambda handlers have the same (large) dependency, the dependency gets compiled separately for each generated handler binary. By contrast, when running something like `go build ... ./cmd/...`, the Go compiler is able to deduplicate common dependencies as it plans out the build process.

While the new `prebuild-lambda` task doesn't directly provide a fix for this issue, it can serve as a useful step for mitigating the performance hit by allowing the Go compiler to cache dependencies in an optimized fashion, which then makes subsequent individual-but-concurrent build processes much faster because the cache is already warmed up.

TL;DR:
- Building many Lambdas individually, but concurrently, can be a drain on system resources when the cache is stale.
- Running a holistic prebuild builds the cache much more optimally, using less system resources.
- As long as the cache is sufficiently warm (e.g. after running the prebuild), the individual-but-concurrent builds executed by Terraform are very fast and don't incur the same drain on system resources.
- Developers can run `task prebuild-lambda` when getting started on development work to take advantage of these speedups.

I also incorporated this utility into the CI/CD workflows. Doing so appears to have cut the build time roughly in half (QA step went from [9m29s](https://github.com/usdigitalresponse/grants-ingest/actions/runs/5339707913) to [4m34s](https://github.com/usdigitalresponse/grants-ingest/actions/runs/5340323962)). 🎉 

## Testing

The net effects of the version changes are best viewed by running `terraform apply` against an environment with the `main` branch checked out, and then running `terraform plan` with this branch checked out. You can get the same experience by checking the plan output embedded in the latest ["Terraform Summary" comment](https://github.com/usdigitalresponse/grants-ingest/pull/144#issuecomment-1601797815) on this PR.

### Automated and Unit Tests
- [X] ~Added Unit tests~ Functionally tested by allowing the "Validate and plan terraform" CI step to run and inspecting the plan output after each version bump.

### Manual tests for Reviewer
- [X] Added steps to test feature/functionality manually

## Checklist
- [x] Provided ticket and description
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers

### CODE REVIEW

**Review by jakekreider — APPROVED** [review]
I tested a plan and apply on a clean workspace and it completed in about 5 minutes from the lambda stage. But note I'm running WSL so I might not be the best test case!

**Comment by TylerHendrickson** [comment]
> I tested a plan and apply on a clean workspace and it completed in about 5 minutes from the lambda stage. But note I'm running WSL so I might not be the best test case!

@jakekreider I'd be interested in hearing a bit more- how does 5 minutes compare to your normal time deploying from an an environment with an empty cache?

**Comment by jakekreider** [comment]
Put a little more rigor into the experimentation this time, running the following on current main and one commit before the merge commit here (`e500fd6b9174e3a8717c8aaf4efbbd6c4b5bcc92~1`)
```
go clean --cache
docker compose down; docker compose up -d
cd terraform
sleep 10 # give time for localstack to make bucket
tflocal init
time tflocal apply -var-file=local.tfvars -auto-approve
cd -
```
`main`:
```
real    2m33.303s
user    6m33.089s
sys     0m39.809s
```

`e500fd6b9174e3a8717c8aaf4efbbd6c4b5bcc92~1` (before this):
```
real    2m49.431s
user    6m16.147s
sys     0m38.172s
```
So pretty comparable. Though one change I made before any of this - somehow my WSL setup was limited to 4GB of memory, and I bumped it to 8
